File: src/tomodachai/game_state.py
"""Central game state manager — single source of truth for a running game session."""

from __future__ import annotations

from datetime import datetime, timezone

from tomodachai.character import Character
from tomodachai.config import AppConfig, load_config
from tomodachai.fountain import FountainManager
from tomodachai.llm import LLMClient
from tomodachai.location import LocationManager
from tomodachai.memory import MemoryStore
from tomodachai.news import NewsManager
from tomodachai.personality import PersonalityType, load_personalities
from tomodachai.relationship import RelationshipTracker
from tomodachai.shop import ShopManager
from tomodachai.simulation import Simulation
from tomodachai.time_system import GameClock


class GameState:
    """Holds everything needed for one game session."""

    def __init__(
        self,
        config: AppConfig | None = None,
        island_name: str = "우리 마을",
        day_count: int = 0,
        money: int = 0,
        time_flip: bool = False,
    ):
        self.config = config or load_config()
        self.personalities: dict[str, PersonalityType] = load_personalities()
        self.characters: list[Character] = []
        self.llm = LLMClient(self.config.llm)
        self._simulation: Simulation | None = None

        # Game-level state
        self.island_name: str = island_name
        self.day_count: int = day_count
        self.money: int = money
        self.time_flip: bool = time_flip
        self.ending_credit_seen: bool = False
        # Global item acquisition catalog (persisted in game.json).
        # Keys: food, clothing, interior, treasure — values: sorted list of item IDs.
        # ShopManager reads from this; do NOT maintain a separate catalog in ShopManager.
        self.catalog: dict[str, list[int]] = {
            "food": [],
            "clothing": [],
            "interior": [],
            "treasure": [],
        }

        # Location system
        self.location_manager: LocationManager = LocationManager()

        # News system
        self.news: NewsManager = NewsManager()

        # Shop system
        self.shop: ShopManager = ShopManager()

        # Fountain events (donation / rap battle / word chain)
        self.fountain: FountainManager = FountainManager()

        # Real-time tracking
        self.last_online: datetime = datetime.now(tz=timezone.utc)
        self._clock: GameClock = GameClock(time_flip=time_flip)

        # 실시간 이벤트 로그 (snapshot since 증분 페치 기준)
        self._event_seq: int = 0
        self._event_log: list[dict] = []

        # 도구 산출물 (세션 한정 — 사진 갤러리 / 요리 카탈로그)
        self.photos: list[dict] = []
        self.dishes: list[dict] = []

    # ------------------------------------------------------------------
    # Character management
    # ------------------------------------------------------------------

    def add_character(self, char: Character) -> Character:
        if any(c.id == char.id for c in self.characters):
            raise ValueError(f"Character with id '{char.id}' already exists")
        self.characters.append(char)
        self._simulation = None  # invalidate
        # 개인 방 등록 (int ID만 지원)
        char_id = char.id if isinstance(char.id, int) else abs(hash(char.id)) % 100000
        self.location_manager.register_private_room(char_id, char.name)
        return char

    def get_character(self, char_id) -> Character | None:
        return next((c for c in self.characters if c.id == char_id), None)

    def remove_character(self, char_id) -> bool:
        before = len(self.characters)
        self.characters = [c for c in self.characters if c.id != char_id]
        if len(self.characters) < before:
            self._simulation = None
            int_id = char_id if isinstance(char_id, int) else abs(hash(char_id)) % 100000
            self.location_manager.remove_character(int_id)
            return True
        return False

    # ------------------------------------------------------------------
    # Economy helpers
    # ------------------------------------------------------------------

    def add_money(self, amount: int) -> None:
        """Add *amount* to the player's money (amount must be positive)."""
        if amount < 0:
            raise ValueError("amount must be non-negative; use spend_money to deduct")
        self.money += amount

    def spend_money(self, amount: int) -> bool:
        """Deduct *amount* from money. Returns True on success, False if insufficient funds."""
        if amount < 0:
            raise ValueError("amount must be non-negative")
        if self.money < amount:
            return False
        self.money -= amount
        return True

    # ------------------------------------------------------------------
    # Catalog helpers (game.json §1)
    # ------------------------------------------------------------------

    _EVENT_LOG_CAP: int = 300  # 실시간 이벤트 로그 ring buffer 상한

    _CATALOG_CATEGORIES = frozenset({"food", "clothing", "interior", "treasure"})

    def add_to_catalog(self, category: str, item_id: int) -> None:
        """Record *item_id* as acquired under *category*. No-op if already present."""
        if category not in self._CATALOG_CATEGORIES:
            valid = sorted(self._CATALOG_CATEGORIES)
            raise ValueError(f"Unknown catalog category '{category}'. Must be one of: {valid}")
        if item_id not in self.catalog[category]:
            self.catalog[category].append(item_id)

    def is_in_catalog(self, category: str, item_id: int) -> bool:
        """Return True if *item_id* has been acquired in *category* before."""
        if category not in self._CATALOG_CATEGORIES:
            valid = sorted(self._CATALOG_CATEGORIES)
            raise ValueError(f"Unknown catalog category '{category}'. Must be one of: {valid}")
        return item_id in self.catalog[category]

    # ------------------------------------------------------------------
    # Simulation access
    # ------------------------------------------------------------------

    @property
    def simulation(self) -> Simulation:
        if self._simulation is None:
            self._simulation = Simulation(
                config=self.config,
                characters=self.characters,
                llm=self.llm,
                personalities=self.personalities,
            )
        return self._simulation

    @property
    def relationships(self) -> RelationshipTracker:
        return self.simulation.relationships

    @property
    def memory(self) -> MemoryStore:
        return self.simulation.memory

    @property
    def bubbles(self) -> list:
        """플레이어 응답 대기 말풍선 큐 (시뮬 소유)."""
        return self.simulation.bubbles

    # ------------------------------------------------------------------
    # Real-time interface
    # ------------------------------------------------------------------

    def _clock_str(self) -> str:
        now = self._clock.now()
        hour = self._clock.get_game_hour(now)
        return f"{hour:02d}:{now.minute:02d}"

    def record_events(self, raw_events: list[dict]) -> None:
        """raw 이벤트들에 단조 seq + 기록 시점 day/clock을 붙여 로그에 적재."""
        clock = self._clock_str()
        for raw in raw_events:
            self._event_seq += 1
            self._event_log.append(
                {"seq": self._event_seq, "day": self.day_count, "clock": clock, "raw": raw}
            )
        if len(self._event_log) > self._EVENT_LOG_CAP:
            self._event_log = self._event_log[-self._EVENT_LOG_CAP :]

    def events_since(self, since: int) -> list[dict]:
        """seq > since 인 로그 항목만 (시간순)."""
        return [e for e in self._event_log if e["seq"] > since]

    def clear_hungry_bubble(self, char_id) -> None:
        """해당 캐릭터의 배고픔 말풍선 제거 (feed 시 호출)."""
        self.bubbles[:] = [
            b for b in self.bubbles if not (b.kind == "hungry" and b.char_id == char_id)
        ]

    def answer_bubble(self, index: int, char_name: str, allow: bool) -> dict:
        """프론트 말풍선 응답 처리.

        반환: {"error": ...} | {"message": ...} | {"scene": ..., "messages": [...]}.
        """
        bubbles = self.bubbles
        if not (0 <= index < len(bubbles)):
            return {"error": "이미 처리된 말풍선입니다"}
        bubble = bubbles[index]
        owner = self.get_character(bubble.char_id)
        if owner is None or owner.name != char_name:
            return {"error": "말풍선이 갱신되었습니다. 다시 시도해 주세요"}
        if bubble.kind == "hungry":
            return {"error": "배고픔 말풍선은 밥을 주면 사라져요"}
        bubbles.pop(index)
        if bubble.kind != "confess_request":
            return {"message": "말풍선을 확인했습니다"}
        result = self.simulation.resolve_confession(bubble, approved=allow)
        self.record_events([result])
        return {"scene": result["summary"], "messages": []}

    def reset_world(self) -> None:
        """새 마을 시작: 캐릭터/시뮬/이벤트로그/seq 초기화 (config는 유지)."""
        self.characters.clear()
        self._simulation = None
        self._event_seq = 0
        self._event_log = []
        self.day_count = 0
        self.photos = []
        self.dishes = []

    def step(self) -> list[dict]:
        """실시간 단일 이벤트 스텝. 서버 백그라운드 태스크에서 호출."""
        if not self.characters:
            return []
        self.touch()
        events = self.simulation.step()
        self.record_events(events)
        return events

    def touch(self) -> None:
        """접속 시각 갱신 (플레이어가 온라인임을 표시)."""
        self.last_online = datetime.now(tz=timezone.utc)

    def connect(self) -> dict:
        """재접속 처리: 오프라인 기간 계산 후 catch-up 이벤트 생성.

        Returns:
            {
                "offline_hours": float,
                "catchup_events": list[dict],
                "time_period": str,
                "is_new_day": bool,
            }
        """
        now_utc = datetime.now(tz=timezone.utc)
        clock = self._clock

        offline_duration = clock.calculate_offline_duration(self.last_online)
        offline_hours = offline_duration.total_seconds() / 3600.0
        is_new_day = clock.is_new_day(self.last_online)

        catchup_events: list[dict] = []
        if offline_hours >= 5.0 / 60.0 and self.characters:  # 최소 5분 이상
            catchup_events = self.simulation.generate_catchup_events(offline_hours)
            if is_new_day:
                self.day_count += 1

        # 일일 리셋 처리 (상점 등 — 현재는 day_count만)
        self.last_online = now_utc

        return {
            "offline_hours": round(offline_hours, 2),
            "catchup_events": catchup_events,
            "time_period": clock.get_time_period(),
            "is_new_day": is_new_day,
        }

    @property
    def current_time_period(self) -> str:
        return self._clock.get_time_period()

    @property
    def current_game_hour(self) -> int:
        return self._clock.get_game_hour()

File: prototype/web/src/sim/gameState.ts
// Python src/tomodachai/game_state.py 미러 — 한 게임 세션의 단일 진실원.
//
// 의존성(LlmClient/SimRng/nowFn)은 주입형 seam. Python은 GameState 내부에서
// LLMClient(config.llm)를 직접 생성하지만, sim 모듈을 프레임워크 무의존으로 유지하기
// 위해 클라이언트(Ollama LlmClient)와 RNG, 시각을 외부에서 주입한다.

import type { AppConfig } from "./config";
import { makeDefaultConfig } from "./config";
import type { Character } from "./character";
import type { LlmClient } from "../llm";
import type { SimRng } from "./rng";
import type { PersonalityType } from "./personalityType";
import { loadPersonalities } from "./personalityType";
import { LocationManager } from "./location";
import { NewsManager } from "./news";
import { ShopManager } from "./shop";
import { FountainManager } from "./fountain";
import { GameClock } from "./clock";
import { Simulation } from "./simulation";
import type { SimEvent, CatchupEvent } from "./simulation";
import type { RelationshipTracker } from "./relationshipTracker";
import type { MemoryStore } from "./memory";

export interface Catalog {
  food: number[];
  clothing: number[];
  interior: number[];
  treasure: number[];
}
export type CatalogCategory = keyof Catalog;

export interface ConnectResult {
  offline_hours: number;
  catchup_events: CatchupEvent[];
  time_period: string;
  is_new_day: boolean;
}

export interface GameStateDeps {
  llm: LlmClient;
  rng: SimRng;
  nowFn?: () => Date;
}

export interface GameStateInit {
  config?: AppConfig;
  islandName?: string;
  dayCount?: number;
  money?: number;
  timeFlip?: boolean;
}

// game.json §1 카탈로그 카테고리 (frozenset 미러)
const CATALOG_CATEGORIES = new Set<string>(["food", "clothing", "interior", "treasure"]);
// Python f-string이 sorted(list)를 그대로 repr하므로 에러 메시지도 동일 포맷 유지.
const CATEGORY_ERR_LIST = "['clothing', 'food', 'interior', 'treasure']";

/** Python round(x, 2) 근사 — 실시간 차이값 파생이라 비트일치 비보장(허용). */
function round2(x: number): number {
  return Math.round(x * 100) / 100;
}

/**
 * 개인 방 등록용 int ID. Python: `id if isinstance(id, int) else abs(hash(id)) % 100000`.
 * Python hash()는 실행마다 salt가 달라 str 경로 자체가 비결정적 → 이식 불가.
 * 게임은 int id만 쓰므로(게임_state.py 주석 "int ID만 지원") number 경로만 충실 포팅,
 * str은 결정론 폴백만 제공(Python과 값 불일치 허용).
 */
function privateRoomId(id: number | string): number {
  if (typeof id === "number") return id;
  let h = 0;
  for (let i = 0; i < id.length; i++) h = (h * 31 + id.charCodeAt(i)) | 0;
  return Math.abs(h) % 100000;
}

export class GameState {
  config: AppConfig;
  readonly personalities: Record<string, PersonalityType>;
  characters: Character[] = [];

  // 게임 레벨 상태 (save 직렬화 대상 — snake_case 유지)
  island_name: string;
  day_count: number;
  money: number;
  time_flip: boolean;
  ending_credit_seen = false;
  catalog: Catalog = { food: [], clothing: [], interior: [], treasure: [] };

  readonly location_manager: LocationManager;
  readonly news: NewsManager;
  readonly shop: ShopManager;
  readonly fountain: FountainManager;
  last_online: Date;

  private readonly llm: LlmClient;
  private readonly rng: SimRng;
  private readonly _clock: GameClock;
  private _simulation: Simulation | null = null;

  constructor(deps: GameStateDeps, init: GameStateInit = {}) {
    this.config = init.config ?? makeDefaultConfig();
    this.personalities = loadPersonalities();
    this.island_name = init.islandName ?? "우리 마을";
    this.day_count = init.dayCount ?? 0;
    this.money = init.money ?? 0;
    this.time_flip = init.timeFlip ?? false;

    this.llm = deps.llm;
    this.rng = deps.rng;

    this.location_manager = new LocationManager();
    this.news = new NewsManager();
    this.shop = new ShopManager();
    this.fountain = new FountainManager();

    this._clock = new GameClock(this.time_flip, deps.nowFn);
    this.last_online = this._clock.now();
  }

  // ------------------------------------------------------------------
  // 캐릭터 관리
  // ------------------------------------------------------------------

  addCharacter(char: Character): Character {
    if (this.characters.some((c) => c.id === char.id)) {
      throw new Error(`Character with id '${char.id}' already exists`);
    }
    this.characters.push(char);
    this._simulation = null; // 무효화
    this.location_manager.registerPrivateRoom(privateRoomId(char.id), char.profile.name);
    return char;
  }

  getCharacter(charId: number | string): Character | null {
    return this.characters.find((c) => c.id === charId) ?? null;
  }

  removeCharacter(charId: number | string): boolean {
    const before = this.characters.length;
    this.characters = this.characters.filter((c) => c.id !== charId);
    if (this.characters.length < before) {
      this._simulation = null;
      this.location_manager.removeCharacter(privateRoomId(charId));
      return true;
    }
    return false;
  }

  // ------------------------------------------------------------------
  // 경제 헬퍼
  // ------------------------------------------------------------------

  /** 양수 금액 추가. 음수면 throw (차감은 spendMoney). */
  addMoney(amount: number): void {
    if (amount < 0) {
      throw new Error("amount must be non-negative; use spend_money to deduct");
    }
    this.money += amount;
  }

  /** 금액 차감. 성공 시 true, 잔액 부족이면 false. 음수면 throw. */
  spendMoney(amount: number): boolean {
    if (amount < 0) {
      throw new Error("amount must be non-negative");
    }
    if (this.money < amount) {
      return false;
    }
    this.money -= amount;
    return true;
  }

  // ------------------------------------------------------------------
  // 카탈로그 헬퍼 (game.json §1)
  // ------------------------------------------------------------------

  /** category 아래에 item_id를 획득 기록. 이미 있으면 무시. */
  addToCatalog(category: string, itemId: number): void {
    if (!CATALOG_CATEGORIES.has(category)) {
      throw new Error(`Unknown catalog category '${category}'. Must be one of: ${CATEGORY_ERR_LIST}`);
    }
    const list = this.catalog[category as CatalogCategory];
    if (!list.includes(itemId)) {
      list.push(itemId);
    }
  }

  /** category에서 item_id를 전에 획득했으면 true. */
  isInCatalog(category: string, itemId: number): boolean {
    if (!CATALOG_CATEGORIES.has(category)) {
      throw new Error(`Unknown catalog category '${category}'. Must be one of: ${CATEGORY_ERR_LIST}`);
    }
    return this.catalog[category as CatalogCategory].includes(itemId);
  }

  // ------------------------------------------------------------------
  // 시뮬레이션 접근
  // ------------------------------------------------------------------

  get simulation(): Simulation {
    if (this._simulation === null) {
      this._simulation = new Simulation(
        this.config,
        this.characters,
        this.llm,
        this.personalities,
        this.rng,
      );
    }
    return this._simulation;
  }

  get relationships(): RelationshipTracker {
    return this.simulation.relationships;
  }

  get memory(): MemoryStore {
    return this.simulation.memory;
  }

  // ------------------------------------------------------------------
  // 실시간 인터페이스
  // ------------------------------------------------------------------

  /** 실시간 단일 이벤트 스텝. */
  async step(): Promise<SimEvent[]> {
    if (this.characters.length === 0) return [];
    this.touch();
    return this.simulation.step();
  }

  /** 접속 시각 갱신. */
  touch(): void {
    this.last_online = this._clock.now();
  }

  /** 재접속 처리: 오프라인 기간 계산 후 catch-up 이벤트 생성. */
  connect(): ConnectResult {
    const nowUtc = this._clock.now();
    const offlineHours = (nowUtc.getTime() - this.last_online.getTime()) / 3_600_000;
    const isNewDay = this._clock.isNewDay(this.last_online);

    let catchupEvents: CatchupEvent[] = [];
    if (offlineHours >= 5.0 / 60.0 && this.characters.length > 0) {
      // 최소 5분 이상
      catchupEvents = this.simulation.generateCatchupEvents(offlineHours);
      if (isNewDay) {
        this.day_count += 1;
      }
    }

    this.last_online = nowUtc;

    return {
      offline_hours: round2(offlineHours),
      catchup_events: catchupEvents,
      time_period: this._clock.getTimePeriod(),
      is_new_day: isNewDay,
    };
  }

  get currentTimePeriod(): string {
    return this._clock.getTimePeriod();
  }

  get currentGameHour(): number {
    return this._clock.getGameHour();
  }
}
